### ISSUE TITLE
terminal bells all over the place using PuTTY

### ISSUE THREAD

**Issue by elandorr** [neovim/neovim#18036]
### Neovim version (nvim -v)

0.4.4

### Vim (not Nvim) behaves the same?

no, 8.2.2434

### Operating system/version

win 10

### Terminal name/version

putty/motty .76

### $TERM environment variable

xterm-256color

### Installation

debian repos

### How to reproduce the issue

just start nvim and try using it

### Expected behavior

bell only on events that matter

### Actual behavior

bells all over the place
the reason for this is here #12769

I'm creating this issue as per @dundargoc's suggestion, it's not urgent, and it's kinda only a follow up on the previous issue: #12769 

The proposed solution of using putty-256color instead doesn't work, as it causes a load of other issues. Apparently most people just don't care or turn them off, but you can find references hidden deep within comments: https://unix.stackexchange.com/questions/20257/how-do-you-display-xterm-colors-with-putty-bash#comment1076855_358608
Another example I just noticed: in nnn it causes colors to shift depending on the direction you move - apparently previous colors affect future ones. It's unreadable even right after opening too.

Even their own docs recommend xterm: https://the.earth.li/~sgtatham/putty/0.76/htmldoc/Chapter4.html#config-colours

In short, the terminfo for putty seems to be widely deprecated/unmaintained/never used in the first place causing all kinds of issues. As there is no other real usable terminal on windows, it has to be putty + xterm-256color. Kitty behaves the same. (The putty fork on win, not the linux bloatware.)

vim doesn't show these excess bells, but it still rings on an ordinary close/save. Is that supposed to happen? I haven't had audio on on my recent vim-adventures. I don't recall it ever doing that in the past either though.

I only came to nvim as there are bugs with colors in regular vim and I wanted to see if it may be fixed in nvim. Same issues though, seems to be related to https://github.com/morhetz/gruvbox/issues/382. Colors are off just enough to be unusable, inversion doesn't work etc. Other applications and test scripts work just fine. I guess not enough people use this 'excessive' stuff and stick with all defaults. On bloatware it works, but using hundreds if not thousands of MiB (looking at the recent 'trend terminals') of RAM is not an option. On windows there is little choice to begin with, and on *nix most is broken, too. I'd like a working combo of putty+xterm/urxvt or if it has to be bigger, lxt, but it seems some stuff is always broken. I'll look for a simpler theme and turn off bells for now. I would like bells for actual errors, though.

Welcome to terminal hell. 20 years later and it's still a mess.

**Comment by clason** [comment]
Neovim 0.4.4 is ancient. Please test with the latest release (0.6.1), or better yet, the current nightly (in which case try to reproduce with `nvim --clean`).

**Comment by elandorr** [comment]
You have a peculiar definition of ancient: 
>  -- James McCoy <jamessan@debian.org>  Wed, 05 Aug 2020 21:06:14 -0400

It's the ordinary stable release.

I might compile myself later! This seems to be a general issue though and not even neovim specific. Compatibility across terminals has always been terrible.

**Comment by clason** [comment]
> It's the ordinary stable release.

Debian has a peculiar definition of "stable" ;)

From the point of Neovim (which is all I care about as maintainer), 0.4.4 _is_ ancient as there have been several releases with a huge number of (relevant) changes in the meantime.

**Comment by elandorr** [comment]
To keep records:

It seems the gruvbox theme was causing at least some trouble. Now with a blank slate (save for some linenumbers + termguicolors) on the 'old' versions:

- vim
  - plain putty/motty xterm-256color
    - no bell on close
    - can trigger bell by hitting esc repeatedly
    - no bell on :q write warning (might have always been like that)
  - inside tmux (same xterm-256color)
    - **bell on close**
    - can trigger bell by hitting esc repeatedly
    - no bell on :q write warning (might have always been like that)
- nvim
  - plain putty/motty xterm-256color
    - bell on open
    - bell on `:`
    - bell on esc to cancel
    - can not trigger bell by hitting esc repeatedly
    - bell on :q write warning (but bells almost everywhere so..)
  - inside tmux (same xterm-256color)
    - no bell on open
    - no bell anywhere as far as I can tell

I'll try to do the same test off master later. For now I've wasted far too much time already :) .

**Comment by dundargoc** [comment]
Wow, thank you for the (very) detailed write-up! I obviously can't make any promises if/when this'll be solved, but at least it has a much higher chance now than before :)

**Comment by elandorr** [comment]
No need to hurry. Terminal compat issues existed for decades. With 'mainstream youtube setups' (didn't know this became the new fidget spinner) it could easily be avoided, but we don't have infinite resources. Your standard ubuntu with gnome now uses more res than a plain win box. One of my recently tested terminals alone even uses >1GiB RAM, and >300MiB storage. Plus half the GPU, for a single blinking block. And it doesn't even provide any more value than xterm 20y ago! I'm surprised that in 2022 basic stuff still isn't fixed universally. A few colors, standard bells, and text formatting in times of hyperrealistic 3D seem trivial. 

Full theme support + regular bells would be a nice bonus, but not critical. At least putty and mintty leave some res for.. working. :P I haven't tried it on urxvt yet, but the bigger issue is a win setup as there's no alternative. 

Even the new 'micro' has the same problem with the cursor by the way, even in mintty.

I already appreciate the little things in vanilla nvim, thank you very much for your efforts! :)

(added later)

I've seriously given the 'trend terminals' alacritty and wez a chance. Ala is better, but still easily used 300-400MiB after a quick test session (with devs calling these leaks 'not an issue'), but:
The cursor is broken even there. I cannot find a terminal where `cursor cterm` works. Not even `printf '%b' '\e]12;red\a'` works in alacritty. Weirdly enough the predestined cterm didn't work for regular colors either, but gui did.
I can't test the bell there as the visual one's animation is bugged and there's no audible.

For an 'nvim standard setup' I'd hope for mintty and/or putty support (that would allow moba, which is very productive, although no italics). Seems to be the best ones standards wise. (And tremendously faster and lighter than ala et al.. They boast with speed, but that speed is wholly imaginary when it blows up resource usage after 5mins of testing.)

Which terminals do you use yourself as a maintainer? The only terminal I've heard of fully working is iTerm2 on Mac. Can't find one for windows even ignoring resource constraints. On linux I've seen gnome's, but the direction gnome has taken doesn't allow for lightweight use anymore. And it has no visual bell. Software gets more and more bloated, yet xterm/urxvt are more productive.

How hard can this be really: `hi Cursor guifg=NONE guibg=NONE gui=reverse ctermfg=NONE ctermbg=NONE cterm=reverse` + standard xterm bells.

I've only been using vim for quick remote matters in vanilla, and it seems it might have to stay that way with so much stuff breaking left and right. Hours of effort in research weren't worth it in retrospect. Thank you for all your own efforts too and sorry for the long post.

**Comment by dundargoc** [comment]
Neovim maintainers use all types of terminals: kitty, alacritty, iterm2, you name it, and many seem satisfied :) . I'd love to have this discussion at our [discourse](https://neovim.discourse.group/) if you don't mind, since issues tend to become very hard to follow after a while if they're not on topic for the problem at hand.

**Comment by erw7** [comment]
I checked putty's [Change Log](https://www.chiark.greenend.org.uk/~sgtatham/putty/changes.html) and could not find any mention of compatibility fixes with xterm after 0.74 (2020-06-27). Therefore, I believe the solution presented in #12769 is still valid today. There is no need to correct problems caused by incorrect `TERM` settings. Because of that, this will be closed.

Please create an issue regarding the following problem when `TERM` is set to `putty-256color`.

> The proposed solution of using putty-256color instead doesn't work, as it causes a load of other issues. Apparently most people just don't care or turn them off, but you can find references hidden deep within comments: https://unix.stackexchange.com/questions/20257/how-do-you-display-xterm-colors-with-putty-bash#comment1076855_358608
Another example I just noticed: in nnn it causes colors to shift depending on the direction you move - apparently previous colors affect future ones. It's unreadable even right after opening too.

**Comment by cjwatson** [comment]
For the record, this was fixed in PuTTY 0.78: https://git.tartarus.org/?p=simon/putty.git;a=commitdiff;h=de66b0313a
